lib/events/athena/querier.go

```diff
@@ -0,0 +1,496 @@
+	// for sessionID != "". This kind of call is done on RBAC to check if user can access that session.
+	filter := searchEventsFilter{eventTypes: []string{events.SessionEndEvent, events.WindowsDesktopSessionEndEvent}}
+	if cond != nil {
+		condFn, err := utils.ToFieldsCondition(cond)
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+		filter.condition = condFn
+	}
+	return q.searchEvents(context.TODO(), fromUTC, toUTC, limit, order, startKey, filter, sessionID)
+}
+
+func (q *querier) searchEvents(ctx context.Context, fromUTC, toUTC time.Time, limit int,
+	order types.EventOrder, startKey string, filter searchEventsFilter, sessionID string,
+) ([]apievents.AuditEvent, string, error) {
+	if limit <= 0 {
+		limit = defaults.EventsIterationLimit
+	}
+	if limit > defaults.EventsMaxIterationLimit {
+		return nil, "", trace.BadParameter("limit %v exceeds %v", limit, defaults.EventsMaxIterationLimit)
+	}
+
+	var startKeyset *keyset
+	if startKey != "" {
+		var err error
+		startKeyset, err = fromKey(startKey)
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+	}
+
+	query, params := prepareQuery(searchParams{
+		fromUTC:     fromUTC,
+		toUTC:       toUTC,
+		order:       order,
+		limit:       limit,
+		startKeyset: startKeyset,
+		filter:      filter,
+		sessionID:   sessionID,
+	}, q.tablename)
+
+	q.logger.WithField("query", query).
+		WithField("params", params).
+		WithField("startKey", startKey).
+		Debug("Executing on Athena")
+
+	queryId, err := q.startQueryExecution(ctx, query, params)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	if err := q.waitForSuccess(ctx, queryId); err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	output, nextKey, err := q.fetchResults(ctx, queryId, limit, filter)
+	return output, nextKey, trace.Wrap(err)
+}
+
+type searchEventsFilter struct {
+	eventTypes []string
+	condition  utils.FieldsCondition
+}
+
+type queryBuilder struct {
+	builder strings.Builder
+	args    []string
+}
+
+// withTicks wraps string with ticks.
+// string params in athena need to be wrapped by "ticks".
+func withTicks(in string) string {
+	return fmt.Sprintf("'%s'", in)
+}
+
+func sliceWithTicks(ss []string) []string {
+	out := make([]string, 0, len(ss))
+	for _, s := range ss {
+		out = append(out, withTicks(s))
+	}
+	return out
+}
+
+func (q *queryBuilder) Append(s string, args ...string) {
+	q.builder.WriteString(s)
+	q.args = append(q.args, args...)
+}
+
+func (q *queryBuilder) String() string {
+	return q.builder.String()
+}
+
+func (q *queryBuilder) Args() []string {
+	return q.args
+}
+
+type searchParams struct {
+	fromUTC, toUTC time.Time
+	limit          int
+	order          types.EventOrder
+	startKeyset    *keyset
+	filter         searchEventsFilter
+	sessionID      string
+}
+
+func prepareQuery(params searchParams, tablename string) (string, []string) {
+	qb := &queryBuilder{}
+	qb.Append(`SELECT DISTINCT uid, event_time, event_data FROM `)
```

Comment:
I don't think a text template would be appropriate since we don't actually want to template the params into the string, the query string and the params should be passed to athena separately. This query builder was pretty confusing at first, could probably use some comments explaining what it's doing, but I do like how it logically groups each section of the query with the relevant params

Reply:
@rosstimothy I have update comments to make it more clear in https://github.com/gravitational/teleport/commit/092f00eac0dfddbf804a8bf8a1454d930afb7309.

text template won't make it more readable, because we want on output both query with placeholders and execution params. If we go with template we will need to build separately template and separately execution parameters instead of doing it in one place.